### ISSUE TITLE
Restart training model from checkpoint

### ISSUE THREAD

**Issue by aslu98** [facebookresearch/ParlAI#2939]
I'm using Colab and after a while my training for fine-tuning gets interrupted due to full storage, su I need to restart my runtime to empy the storage and then restart the training from the checkpoint, but that does not seem to happen (the ppl is the one there was at the beggining of the training, not at the end of the previous training).
I save my models in Drive so when I restart the session they do not get removed.
The "half-trained" model is not usable, as it produces answers like: `__unk__ __unk__ br __unk__ __unk__ ; __unk__ __unk__ Ġhilarious __unk__ __unk__ Ġpace __unk__ __unk__ Ġshots __unk__ __unk__ PC __unk__ __unk__ Ġteachers __unk__ __unk__ z __unk__ __unk__ Ġalleg __unk__ __unk__ Ġgun __unk__ __unk__ AT __unk__ __unk__ uth __unk__ __unk__ zen __unk__ __unk__ Ġbring __unk__ __unk__ den __unk__ __unk__ ^^ __unk__ __unk__ Ġeconomy __unk__ __unk__ av __unk__ __unk__ zer [...]` to messages like "Hi", so I need to complete training.

I tried to set `--init-model` and `--init-model-dict` to my_model or my_model.checkpoint or just remove these args but it seems like it restarts the training from the beginning everytime.
I also put `--load-from-checkpoint` True, even if it should be the default (#2894).

I'm sorry about the amount of issues I'm opening but I did not find anything about loading model from checkpoint in the documentation, how can I do it?

Also, the fact that I need to restart my runtime so often is not very convenient, why the training uses Colab storage to save temporary files even if my Parlai directory is in the Drive folder? Is there a way to make it save the files it needs for training in the directory I want (Drive)?

**Comment by stephenroller** [comment]
Yeah it sounds like the files are being lost somehow.

When you start the colab notebook, can you `ls` the files and see if they're there? You're using the [drive mounting](https://towardsdatascience.com/downloading-datasets-into-google-drive-via-google-colab-bcb1b30b0166) support? Are you sure you're successfully remounting the drive?

The right way to resume training is make the exact same call to train_model (so --model-file will point to the same place). Our cluster has heavy preemption, so we resume training successfully with great frequency.

**Comment by stephenroller** [comment]
I just tried this, with a factory reset between my runs and it worked quite well:

```
!nvidia-smi
from google.colab import drive
drive.mount("/content/gdrive")
!pip install parlai
!mkdir -p "/content/gdrive/My Drive/parlaitest"


from parlai.scripts.train_model import TrainModel

TrainModel.main(
    model='transformer/generator',
    task='convai2',
    num_epochs=1,
    veps=0.05,
    batchsize=8,
    skip_generation=True,
    vmt='ppl',
    model_file='/content/gdrive/My Drive/parlaitest/model'
)

2020-08-05 11:12:09,749 INFO     | building dictionary first...
2020-08-05 11:12:09,960 INFO     | Using CUDA
2020-08-05 11:12:09,966 INFO     | loading dictionary from /content/gdrive/My Drive/parlaitest/model.dict
2020-08-05 11:12:10,197 INFO     | num words = 18760
2020-08-05 11:12:19,272 INFO     | Total parameters: 9,138,000 (8,523,600 trainable)
2020-08-05 11:12:19,274 INFO     | Loading existing model params from /content/gdrive/My Drive/parlaitest/model
2020-08-05 11:12:20,853 INFO     | creating task(s): convai2
[building data: /usr/local/lib/python3.6/dist-packages/data/ConvAI2]
2020-08-05 11:12:20,859 INFO     | Downloading http://parl.ai/downloads/convai2/convai2_fix_723.tgz to /usr/local/lib/python3.6/dist-packages/data/ConvAI2/convai2_fix_723.tgz
Downloading convai2_fix_723.tgz: 100%|██████████| 430M/430M [00:21<00:00, 20.0MB/s]
2020-08-05 11:12:55,788 INFO     | loading fbdialog data: /usr/local/lib/python3.6/dist-packages/data/ConvAI2/train_self_original.txt
```

Notice the line `2020-08-05 11:17:22,681 INFO     | Loading existing model params from /content/gdrive/My Drive/parlaitest/model`

I let it run for a few minutes and reset (running the above commands again):
```
[  warmup_updates: -1 ]
2020-08-05 11:18:02,780 INFO     | training...
2020-08-05 11:18:03,320 INFO     | time:277s total_exs:26328 epochs:0.20 time_left:1107s
    clip  ctpb  ctps  exps  exs  gnorm  gpu_mem  loss  lr  ltpb  ltps   ppl  token_acc  total_train_updates  tpb  tps   ups
       1   489  1115  17.1    8  4.212    .0152 2.803   1    89 202.8 16.49      .4045                 3291  578 1318 2.281
```

Notices "Epochs" resumes where I left off.

**Comment by aslu98** [comment]
Thank you so much for the quick answers and the example.

I mounted drive and my "File" section looks like this: https://ibb.co/GQhkj3d (my models are in the covidmodels repository).
When I re-start runtime the Colab Storage is about half empty and I need to remount drive, I start the `train_model` inside the drive/My\ Drive/ParlAI directory but the Colab Storage starts increasing until the notebook crashes.
I thought the only files training needs to create would be in the `--model-file` directory.

I noticed that my fine-tuning output actually states `09:37:23 INFO | Loading existing model params from covidmodels/covid3/covid_model3.checkpoint`, but then why the first ppl score of this new fine-tuning "round" is like 8878  whereas the last ppl of the previous "round" was like 223?

**Comment by stephenroller** [comment]
Maybe you’re running out of storage and the write is failing? The model checkpoint + a temporary file we use to ensure writes are atomic mean that the maximum storage peaks at 3x model size. You can reduce that to 2x by disabling check pointing (it’s something like —save-after-valid)

Are you validating often enough? The last validation will be your resumption point.

I suggest trying it with a tiny model to see if its an issue with storage amount.

**Comment by aslu98** [comment]
I was trying with blender90M, I turned —save-after-valid off like you suggested but nothing changed
My output stated something like:
 `09:37:30 INFO | saving model checkpoint: covidmodels/covid3/covid_model3.checkpoint
09:37:31 REPO | did not beat best ppl: 4403.9002 impatience: 1` 

and then:
 `09:37:31 INFO | running eval: valid
/usr/local/lib/python3.6/dist-packages/torch/cuda/memory.py:234: FutureWarning: torch.cuda.reset_max_memory_allocated now calls torch.cuda.reset_peak_memory_stats, which resets /all/ peak memory stats.
  FutureWarning)
09:37:32 INFO | eval completed in 0.30s
09:37:32 REPO | valid:`

more then once before going `No space left on device`, so I guess they where saving the model.

Anyway, I figured out Colab was not the right choice for fine-tuning, I managed to do it with a local GPU and it worked out.
